library/bignum.c

```diff
@@ -1934,10 +1934,13 @@ static void mpi_montmul( mbedtls_mpi *A, const mbedtls_mpi *B, const mbedtls_mpi
         u0 = A->p[i];
         u1 = ( d[0] + u0 * B->p[0] ) * mm;
 
-        mpi_mul_hlp( m, B->p, d, u0 );
-        mpi_mul_hlp( n, N->p, d, u1 );
-
-        d++; d[n + 1] = 0;
+        (void) mpi_mul_hlp( d, n + 2,
+                            B->p, m,
+                            u0 );
+        (void) mpi_mul_hlp( d, n + 2,
+                            N->p, n,
+                            u1 );
+        d++;
     }
 
     /* At this point, d is either the desired result or the desired result
```

Comment:
The conditional subtraction can be simplified here to
```c
    borrow = mpi_sub_hlp( n, d, d, N->p );
    borrow = ( d[n] < borrow );
    (void) mpi_mul_hlp( d, n, N->p, n, borrow );
    memcpy( A->p, d, n * ciL );
```
which would be an example of a call to `mpi_mul_hlp()` where you expect and process the carry.

(@yanesca @mpg we touched on this approach in the chat, so FYI).